File: ios/SessionTimeoutModule.swift
import Foundation
import React

@objc(SessionTimeoutModule)
class SessionTimeoutModule: NSObject {
    private var timer: Timer?
    private var originalTimeout: TimeInterval = 0    // Never changes after startTimer
    private var currentDuration: TimeInterval = 0    // Duration for current scheduled timer
    private var remainingTime: TimeInterval = 0
    private var isActive: Bool = false
    private var isPaused: Bool = false
    private var lastResetTime: Date?
    
    @objc
    static func requiresMainQueueSetup() -> Bool {
        return true
    }
    
    @objc
    func startTimer(_ timeout: Double,
                   resolver: @escaping RCTPromiseResolveBlock,
                   rejecter: @escaping RCTPromiseRejectBlock) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            
            self.originalTimeout = timeout / 1000.0  // Convert ms to seconds
            self.currentDuration = self.originalTimeout
            self.remainingTime = self.originalTimeout
            self.lastResetTime = Date()
            self.isActive = true
            self.isPaused = false
            
            self.scheduleTimeout(duration: self.currentDuration)
            resolver(nil)
        }
    }
    
    @objc
    func stopTimer(_ resolver: @escaping RCTPromiseResolveBlock,
                  rejecter: @escaping RCTPromiseRejectBlock) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            
            self.timer?.invalidate()
            self.timer = nil
            self.isActive = false
            self.isPaused = false
            self.remainingTime = 0
            self.currentDuration = 0
            resolver(nil)
        }
    }
    
    @objc
    func resetTimer(_ resolver: @escaping RCTPromiseResolveBlock,
                   rejecter: @escaping RCTPromiseRejectBlock) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            
            // Don't reset if timer was never started
            guard self.originalTimeout > 0 else {
                resolver(nil)
                return
            }
            
            self.timer?.invalidate()
            self.timer = nil
            
            self.currentDuration = self.originalTimeout
            self.remainingTime = self.originalTimeout
            self.lastResetTime = Date()
            self.isPaused = false
            self.isActive = true
            
            self.scheduleTimeout(duration: self.currentDuration)
            
            resolver(nil)
        }
    }
    
    @objc
    func pauseTimer(_ resolver: @escaping RCTPromiseResolveBlock,
                   rejecter: @escaping RCTPromiseRejectBlock) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            
            if !self.isPaused && self.isActive {
                self.timer?.invalidate()
                self.timer = nil
                
                if let lastReset = self.lastResetTime {
                    let elapsed = Date().timeIntervalSince(lastReset)
                    self.remainingTime = max(0, self.currentDuration - elapsed)
                }
                
                self.isPaused = true
            }
            
            resolver(nil)
        }
    }
    
    @objc
    func resumeTimer(_ resolver: @escaping RCTPromiseResolveBlock,
                    rejecter: @escaping RCTPromiseRejectBlock) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            
            if self.isPaused && self.isActive {
                self.lastResetTime = Date()
                self.currentDuration = self.remainingTime
                self.isPaused = false
                self.scheduleTimeout(duration: self.currentDuration)
            }
            
            resolver(nil)
        }
    }
    
    @objc
    func getRemainingTime(_ resolver: @escaping RCTPromiseResolveBlock,
                         rejecter: @escaping RCTPromiseRejectBlock) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else {
                resolver(0)
                return
            }
            
            if !self.isActive {
                resolver(0)
                return
            }
            
            if self.isPaused {
                resolver(self.remainingTime * 1000.0)  // Convert to ms
            } else {
                if let lastReset = self.lastResetTime {
                    let elapsed = Date().timeIntervalSince(lastReset)
                    let remaining = max(0, self.currentDuration - elapsed)
                    resolver(remaining * 1000.0)  // Convert to ms
                } else {
                    resolver(0)
                }
            }
        }
    }
    
    @objc
    func isTimerActive(_ resolver: @escaping RCTPromiseResolveBlock,
                      rejecter: @escaping RCTPromiseRejectBlock) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else {
                resolver(false)
                return
            }
            resolver(self.isActive && !self.isPaused)
        }
    }
    
    private func scheduleTimeout(duration: TimeInterval) {
        timer?.invalidate()
        timer = nil
        
        timer = Timer.scheduledTimer(withTimeInterval: duration, repeats: false) { [weak self] _ in
            guard let self = self else { return }
            self.isActive = false
            self.remainingTime = 0
        }
    }
    
    @objc
    func addListener(_ eventName: String) {
    }
    
    @objc
    func removeListeners(_ count: Double) {
    }
    
    deinit {
        timer?.invalidate()
        timer = nil
    }
}

File: android/src/main/java/com/sessiontimeout/SessionTimeoutModule.java
package com.sessiontimeout;

import android.os.Handler;
import android.os.Looper;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.LifecycleEventListener;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.module.annotations.ReactModule;

@ReactModule(name = SessionTimeoutModule.NAME)
public class SessionTimeoutModule extends ReactContextBaseJavaModule implements LifecycleEventListener {
    public static final String NAME = "SessionTimeoutModule";
    
    private final ReactApplicationContext reactContext;
    private Handler handler;
    private Runnable timeoutRunnable;
    
    private long originalTimeout = 0;      // Never changes after startTimer
    private long currentDuration = 0;      // Duration for current scheduled timer
    private long remainingTime = 0;
    private boolean isActive = false;
    private boolean isPaused = false;
    private long lastResetTime = 0;

    public SessionTimeoutModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
        this.handler = new Handler(Looper.getMainLooper());
        reactContext.addLifecycleEventListener(this);
    }

    @Override
    @NonNull
    public String getName() {
        return NAME;
    }

    @ReactMethod
    public void startTimer(double timeout, Promise promise) {
        try {
            this.originalTimeout = (long) timeout;
            this.currentDuration = (long) timeout;
            this.remainingTime = (long) timeout;
            this.lastResetTime = System.currentTimeMillis();
            this.isActive = true;
            this.isPaused = false;
            
            scheduleTimeout(currentDuration);
            promise.resolve(null);
        } catch (Exception e) {
            promise.reject("START_TIMER_ERROR", e.getMessage(), e);
        }
    }

    @ReactMethod
    public void stopTimer(Promise promise) {
        try {
            if (timeoutRunnable != null) {
                handler.removeCallbacks(timeoutRunnable);
                timeoutRunnable = null;
            }
            isActive = false;
            isPaused = false;
            remainingTime = 0;
            currentDuration = 0;
            promise.resolve(null);
        } catch (Exception e) {
            promise.reject("STOP_TIMER_ERROR", e.getMessage(), e);
        }
    }

    @ReactMethod
    public void resetTimer(Promise promise) {
        try {
            // Don't reset if timer was never started
            if (this.originalTimeout <= 0) {
                promise.resolve(null);
                return;
            }
            
            if (timeoutRunnable != null) {
                handler.removeCallbacks(timeoutRunnable);
            }
            
            this.currentDuration = this.originalTimeout;
            this.remainingTime = this.originalTimeout;
            this.lastResetTime = System.currentTimeMillis();
            this.isPaused = false;
            this.isActive = true;
            
            scheduleTimeout(currentDuration);
            
            promise.resolve(null);
        } catch (Exception e) {
            promise.reject("RESET_TIMER_ERROR", e.getMessage(), e);
        }
    }

    @ReactMethod
    public void pauseTimer(Promise promise) {
        try {
            if (!isPaused && isActive) {
                if (timeoutRunnable != null) {
                    handler.removeCallbacks(timeoutRunnable);
                }
                
                long elapsed = System.currentTimeMillis() - lastResetTime;
                remainingTime = Math.max(0, currentDuration - elapsed);
                isPaused = true;
            }
            promise.resolve(null);
        } catch (Exception e) {
            promise.reject("PAUSE_TIMER_ERROR", e.getMessage(), e);
        }
    }

    @ReactMethod
    public void resumeTimer(Promise promise) {
        try {
            if (isPaused && isActive) {
                lastResetTime = System.currentTimeMillis();
                currentDuration = remainingTime;
                isPaused = false;
                scheduleTimeout(currentDuration);
            }
            promise.resolve(null);
        } catch (Exception e) {
            promise.reject("RESUME_TIMER_ERROR", e.getMessage(), e);
        }
    }

    @ReactMethod
    public void getRemainingTime(Promise promise) {
        try {
            if (!isActive) {
                promise.resolve(0.0);
                return;
            }
            
            if (isPaused) {
                promise.resolve((double) remainingTime);
            } else {
                long elapsed = System.currentTimeMillis() - lastResetTime;
                long remaining = Math.max(0, currentDuration - elapsed);
                promise.resolve((double) remaining);
            }
        } catch (Exception e) {
            promise.reject("GET_REMAINING_TIME_ERROR", e.getMessage(), e);
        }
    }

    @ReactMethod
    public void isTimerActive(Promise promise) {
        try {
            promise.resolve(isActive && !isPaused);
        } catch (Exception e) {
            promise.reject("IS_TIMER_ACTIVE_ERROR", e.getMessage(), e);
        }
    }

    private void scheduleTimeout(long duration) {
        if (timeoutRunnable != null) {
            handler.removeCallbacks(timeoutRunnable);
        }
        
        timeoutRunnable = new Runnable() {
            @Override
            public void run() {
                isActive = false;
                remainingTime = 0;
            }
        };
        
        handler.postDelayed(timeoutRunnable, duration);
    }

    @Override
    public void onHostResume() {
    }

    @Override
    public void onHostPause() {
    }

    @Override
    public void onHostDestroy() {
        if (timeoutRunnable != null) {
            handler.removeCallbacks(timeoutRunnable);
            timeoutRunnable = null;
        }
    }

    @Override
    public void invalidate() {
        super.invalidate();
        if (timeoutRunnable != null) {
            handler.removeCallbacks(timeoutRunnable);
            timeoutRunnable = null;
        }
        reactContext.removeLifecycleEventListener(this);
    }

    @ReactMethod
    public void addListener(String eventName) {
    }

    @ReactMethod
    public void removeListeners(Integer count) {
    }
}